importccl: parallelize the file upload in export csv

adityamaru opened the pull request:
This change parallelizes the stage of export where we upload
files to external storage. Benchmarks have shown that this
is where we spend the maximum amount of time. Adding parallelization
means that we increase the chances of OOMing since we do not
support streaming writes (yet). To protect against this, the
next commit will add a max file size guard.

Release note: None

adityamaru committed:
importccl: parallelize the file upload in export csv

This change parallelizes the stage of export where we upload
files to external storage. Benchmarks have shown that this
is where we spend the maximum amount of time. Adding parallelization
means that we increase the chances of OOMing since we do not
support streaming writes (yet). To protect against this, the
next commit will add a max file size guard.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64384)
<!-- Reviewable:end -->

adityamaru commented:
I'm still in the process of seeing whether this really helps performance but just wanted to get a review on the synchronization between the various goroutines. I thought about the error handling/teardown of the goroutines for a while so I'm happy to sync offline to explain my reasoning. Let me know if there is a more idiomatic way to structure this!

adityamaru commented on pkg/ccl/importccl/exportcsv.go:
```diff
@@ -263,41 +354,39 @@ func (sp *csvWriter) Run(ctx context.Context) {
 		return nil
 	}()
+	if err != nil {
+		_ = group.Wait()
```
should we be waiting here?

adityamaru committed:
importccl: add a cluster setting to control the target file size to export

This change adds a cluster setting `bulkio.export.target_file_size`
that default to 50MiB and determines the target file size when
of the csv files generated during an export. This settings overrides
the exisiting `chunk_rows` setting that controls how many rows we attempt to
chunk into a single csv file.
This setting is important with the parallelisation of export upload
because now we are buffering 10*fileSize when writing to external storage.
This will become less important once streaming writes are implemented.

Release note (sql change): Cluster setting `bulkio.export.target_file_size`
to control the target csv file size when running an EXPORT.

adityamaru commented:
The results are quite varied, but it seems like it makes a little faster:

Running a 2Tb bank export with chunk_rows=5000 on a 10 node default AWS roachprod cluster led to the following timings:

**With parallelization**

`13025 files, 50Mib each in 1073s`
`13025 files, 50Mib each in 704s`
`13025 files, 50Mib each in 2000s`

**Without parallelization**
`13025 files, 50Mib each in 2130s`
`13025 files, 50Mib each in 3008s`

dt commented on pkg/ccl/importccl/exportcsv.go:
```diff
@@ -168,11 +186,98 @@ func (sp *csvWriter) MustBeStreaming() bool {
+	numWorkers := 10
+	// Leave enough space for the upload workers and the goroutine emitting a row
+	// to buffer errors. We will only use the first when reporting an error.
+	errCh := make(chan error, numWorkers+1)
```
Huh, why can't we just use the group's err?

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64384)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @pbardea)

___
*[pkg/ccl/importccl/exportcsv.go, line 242 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64384#-MZnvoFJ36_wc6yNJ3zQ:-MZnvoFJ36_wc6yNJ3zR:b-1nq5tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e277705b63a48d043d90c83bfe641c7b90988bf/pkg/ccl/importccl/exportcsv.go#L242)):*
> ```Go
> 	defer span.Finish()
> 
> 	numWorkers := 10
> ```

a better default might be:
```
const numWorkers = util.ConstantWithMetamorphicTestRange("num-export-workers", runtime.NumCPU(), 1, 2*runtime.NumCPU))
```
___
*[pkg/ccl/importccl/exportcsv.go, line 255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64384#-MZnv70M2FoTlfJWCUTZ:-MZnv70M2FoTlfJWCUT_:b-bqnexc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e277705b63a48d043d90c83bfe641c7b90988bf/pkg/ccl/importccl/exportcsv.go#L255)):*
> ```Go
> 		defer close(sp.resCh)
> 		return ctxgroup.GroupWorkers(ctx, numWorkers, func(ctx context.Context, _ int) error {
> 			if err := sp.uploadCSVFile(ctx); err != nil {
> ```

somewhat of a nit... but maybe not a nit...
It's a bit hard to follow code when channels, readers, writers, etc are all over the place.
Personally, I would find it easier to read this code if:
  1. This function created all of the channels necessary for orchestration.
      The channels will be created and closed in the same part of the code (this function).
  2. I would either have the the channel loops inside these go routines, calling to a helper function (such as upload) to upload a single item.
      OR 
      Rename helpers to something like `uploadCSVFileWorker`, and pass the channel to the worker function explicitly.
 3. Same for the emitResultRow helper below.
___
*[pkg/ccl/importccl/exportcsv.go, line 249 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64384#-MZT2-7tEwbTDtyM7w-L-r3-249:-MZnuSOz2O36PV-KjNGl:b4ejilx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02486784d92d61d103193ddb035d20b51c7ef40/pkg/ccl/importccl/exportcsv.go#L249)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Huh, why can't we just use the group's err?
</blockquote></details>

+ 1 .. we should use group err.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64384)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @pbardea)

___
*[pkg/ccl/importccl/exportcsv.go, line 249 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64384#-MZT2-7tEwbTDtyM7w-L-r3-249:-MZnwKorBjqUA12quzgn:b-oy6sv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02486784d92d61d103193ddb035d20b51c7ef40/pkg/ccl/importccl/exportcsv.go#L249)):*
> e should use group err.
I meant to add my vote for @dt suggestion.


<!-- Sent from Reviewable.io -->

dt commented:
for the record, we talked about this in bulk's weekly -- we want to see if just doing smaller, size-listed files is enough of a win for backports, and then parallelization might wait for streaming uploads to reduce the memory overhead (if it is even still required at that point). The constant `10` is also probably not ideal, and we instead want to start new upload workers by reserving the amount of memory they'll need for their buffering from a monitor, rather than a fixed number. Even then, I'd guess we limit it to something like 4, particularly once we do streaming uploads since I doubt we'll be spending as much time on IO.

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64384)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @pbardea)

___
*[pkg/ccl/importccl/exportcsv.go, line 249 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64384#-MZT2-7tEwbTDtyM7w-L-r3-249:-MZqOaeS3j1ERMS9KSNT:b-e3u1qp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02486784d92d61d103193ddb035d20b51c7ef40/pkg/ccl/importccl/exportcsv.go#L249)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> e should use group err.
I meant to add my vote for @dt suggestion.
</blockquote></details>

My understanding was that the group's error is only returned once all the goroutines have completed? This errCh is so that we can break out of the loop that is performing the datum->csv conversion if any one of the workers fails to upload. We probably don't want to Wait() (to get the error) for all the workers to finish uploading on every iteration of the csv file producing gorotuine? I do use the groups error right at the end when we wait for all workers to complete. I might be missing something...


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64384)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @miretskiy, and @pbardea)

___
*[pkg/ccl/importccl/exportcsv.go, line 249 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64384#-MZT2-7tEwbTDtyM7w-L-r3-249:-MZrJqTP5FYsXU3awyU0:bwpknqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02486784d92d61d103193ddb035d20b51c7ef40/pkg/ccl/importccl/exportcsv.go#L249)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

My understanding was that the group's error is only returned once all the goroutines have completed? This errCh is so that we can break out of the loop that is performing the datum->csv conversion if any one of the workers fails to upload. We probably don't want to Wait() (to get the error) for all the workers to finish uploading on every iteration of the csv file producing gorotuine? I do use the groups error right at the end when we wait for all workers to complete. I might be missing something...
</blockquote></details>

as soon as one worker returns an error, the ctx in the rest is cancelled. We _should_ be waiting for them to exit before we return, as otherwise that's a goroutine leak, but we shouldn't need to wait long since we cancelled the ctx as soon as the first error returned.


<!-- Sent from Reviewable.io -->

shermanCRL commented:
Are we still pursuing this approach? Maybe it’s been replaced by @adityamaru ’s SQL script to chunk the export?

adityamaru commented:
I don't think we're actively pursuing it at the moment, but with streaming writes + memory monitoring it might be worth seeing if there are significant perf gains.

shermanCRL commented:
Is this worth re-opening? Have we gained streaming writes (and memory monitoring) since this was last considered?
